include/IECoreImage/DisplayDriverServer.h

```diff
@@ -66,6 +66,13 @@ class IECOREIMAGE_API DisplayDriverServer : public IECore::RunTimeTyped
 		DisplayDriverServer( int portNumber = 0 );
 		~DisplayDriverServer() override;
 
+		/// Used to artificially limit the available ports.
+		/// Automated port selection (via `portNumber = 0`) will
+		/// be clamped to this range and manually specifying an
+		/// out-of-range port will throw InvalidArgumentException
+		static void setPortRange( const Imath::V2i &range );
```

Comment:
Using `V2i` is definitely a bit strange in that there's nothing geometric about the range. And it also provides a far greater range of values than allowable as a port number. There's a tiny bit of precedent in that we use `V2f` for the `clippingPlanes` in the Camera, but there we have less issues with the range, and we are forced to get it into a CompoundData somehow.

I guess the closest match is probably `std::pair<uint16_t`>? What's the objection to that? More boilerplate to convert to/from a tuple in the bindings?

I suppose another possibility inspired by your suggestion of `FrameRange` is just to pass a string in our standard `1-10, 100-110` format and then parse that into an arbitrary list?

Reply:
Thanks for the feedback, I've tried to reply to each point (though not in the originally written order)

> a string in our standard 1-10, 100-110 format and then parse that into an arbitrary list?

Do we have/envision a need for an arbitrary list? I'm not much of a sockets person, so maybe it'd be useful in some scenario I don't see right now.

> int provides a far greater range of values than allowable as a port number, I guess the closest match is probably uint16_t

Assuming we're not also changing `DisplayDriverServer( int portNumber = 0 );` and `int portNumber()` (which we're not as I need this on Cortex 10.1) I'd find it odd to have `setPortRange` use `uint16_t` rather than `int`.

>  Using V2i is definitely a bit strange in that there's nothing geometric about the range.

For me, non-geometric-ness doesn't particularly matter. Without getting into debate about the various issues with `GeometricTypedData`, there are several other places we use `V2i` for non-geometric purposes, so I wouldn't rule it out on those grounds. At IE at least, we use `V2i` as the defacto `int[2]` type in Cortex.

I mainly questioned it because I think `setPortRange( int min, int max )` would be the nicest setter, but I was worried changing types for the getter could be awkward.

> std::pair? More boilerplate to convert to/from a tuple in the bindings?

Personally I find pairs/tuples annoying in both c++ and python. But let's try it out:

```
static void setPortRange( const std::pair<int, int> &range );
static std::pair<int, int> getPortRange(); // should this be const & instead?
```

use in c++
```
DisplayDriverServer::setPortRange( { 1, 10 } ); // fine
auto portRange = DisplayDriverServer::getPortRange(); // fine
for( auto p = portRange->first; p < portRange->second + 1, ++p ) // awkward, though I could get over it...
// or
int minPort, maxPort;
std::tie( minPort, maxPort ) = DisplayDriverServer::getPortRange(); // horrendous
for( auto p = minPort; p < maxPort + 1, ++p ) // nice
```

use in python
```
IECoreImage.DisplayDriverServer.setPortRange( ( 1, 10 ) ) # awkward, the user needs to guess the data format
portRange = IECoreImage.DisplayDriverServer.getPortRange() # fine
for p in range( portRange[0], portRange[1] + 1 ) # fine
# or
minPort, maxPort = IECoreImage.DisplayDriverServer.getPortRange() # kinda nice
for p in range( minPort, maxPort + 1 ) # nice
```

I'm not sure where that leaves me... Would differing types (ie. `set( int, int )` & `pair<int, int> get()`) be too weird? That would solve the parts I find awkward in python and I could live with the slight annoyance of `.first`, `.second` in c++

Reply:
> Do we have/envision a need for an arbitrary list?

I don't, no. I was just throwing out ideas that avoided some of the awkwardness of `V2i`. I guess another would be `vector<int>` or `vector<uint16_t>`. How wide is the range that you envision using? Too big to want to specify explicitly like that?

> Assuming we're not also changing DisplayDriverServer( int portNumber = 0 ); and int portNumber() (which we're not as I need this on Cortex 10.1) I'd find it odd to have setPortRange use uint16_t rather than int.

Consistency with previous mistakes isn't a particularly strong argument. I double checked and the underlying boost code we call takes `unsigned short`, so the range provided by `int` definitely allows people to ask for the impossible, and we have no checks in place. Choosing types that enforce constraints automatically is generally a pretty good thing.

> for( auto p = portRange->first; p < portRange->second + 1, ++p ) // awkward, though I could get over it...

I don't really see a problem there.

> IECoreImage.DisplayDriverServer.setPortRange( ( 1, 10 ) ) # awkward, the user needs to guess the data format

My first guess would be either `tuple( min, max )` or just `min, max`. Finding a vector type here would be a surprise.

> I'm not sure where that leaves me...

It leaves me arguing for principles I do care about, but in a function that probably isn't that important in itself :)

Let's summarise :

- `int` is bad, with the vast majority of representable values being completely invalid for our use case. `uint16_t` is the right type.
- `uint16_t, uint16_t` : Convenient to set, convenient to use. Lack of symmetry in get. No natural enforcement of `min <= max`. Only a single range.
- `pair<uint16_t, uint16_t>`, tuple in Python : Symmetry, possibly slightly less convenient, but not by much. No natural enforcement of `min <= max`. Only a single range.
- `V2i` : Symmetry. Surprising use of geometric type. Wrong integer type. No natural enforment of `min <= max`. Only a single range.
- `string` (1-10 etc) : Symmetry. Arbitrary unions of ranges. Compact representation. Lots of invalid (unparseable) values. Overkill and less convenient for the simple case.
- `vector<uint16_t>` : Symmetry. Arbitrary unions of ranges. No invalid values. Wasteful representation. Overkill and less convenient for the simple case.

I'll merge whatever you choose, but for me `pair` or `uint16_t. uint16_t` is probably the winner.

Reply:
> Consistency with previous mistakes isn't a particularly strong argument... Choosing types that enforce constraints automatically is generally a pretty good thing.

I completely agree, but I think a class with so few public methods, all todo with a single concept, should use the same data type in each of those methods. I've gone with `int` for 10.1 in bf9cc66baf2794907048bb6addb3da24b07c609e and switched them all to `uint16_t` in 10.2 via ee48cf74f4459b65b7aa0b31a19928f075a93f5c

Note in bf9cc66baf2794907048bb6addb3da24b07c609e I've also added a registration mechanism. It turns out when using this in practice we need to know the range available in Nuke while we're running in Gaffer. With the registry in place, I think I prefer the symmetry of the pair over the individual `min, max` args, so that the client code is `setPortRange( registeredPortRange( appName ) )`